### PR TITLE
#5562: resolve reduce scatter issues (nd hang and correctness)

### PR DESCRIPTION
**Pull request by SeanNijjar**
### Ticket
- [Link to Github Issue.Link to Github Issue.](https://github.com/tenstorrent/tt-metal/issues/5562)

### Problem description
- Reduce scatter had some general instability the correctness issues to be resolved before being enabled in testing

### What's changed
- Fix bug with reduce scatter reader with its iteration space. Specifically, an issue with computing the page offset in the storage when looping to the next worker slice offset within the tensor slice

- Patch bug with host ccl tensor slicer that couldn't find a valid, non-zero sized worker slice. In such cases, choose a valid and safe slice shape (1x1) as a fallback - that is slice that is 1 page high by one page wide, *not* one element .

- Fix edm builder builder bug with handshake region size in L1
  - Handshake region needs 3 * 16B

This commit also enables reduce scatter testing in post_commit and nightly

### Checklist
- [x] Post commit CI passes
  - post_commit: https://github.com/tenstorrent/tt-metal/actions/runs/9685442045
  - t3000 frequent tests: https://github.com/tenstorrent/tt-metal/actions/runs/9672662627
- [x] Model regression CI testing passes (if applicable)
  - N/A - this op was stable for use yet  
- [x] New/Existing tests provide coverage for changes
   - This commit enables reduce scatter testing in post_commit and nightly. Some additional gtest cases were also addedd

### CODE REVIEW

File: tests/scripts/run_tests.sh [hunk 1]
```diff
@@ -82,6 +82,7 @@ run_frequent_api_pipeline_tests() {
     else
         if [[ $tt_arch == "wormhole_b0" ]]; then
             pytest  tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k nightly
+            pytest  tests/tt_eager/python_api_testing/unit_testing/misc/test_reduce_scatter.py -k nightly
```
Comment (tt-rkim) [inline tests/scripts/run_tests.sh:85]: Is it possible you can factor out the nightly tests into separate files?

We want to reduce usage of `-k` per #7586

Comment (SeanNijjar) [reply tests/scripts/run_tests.sh:85]: Hey @tt-rkim, are you suggesting here that I make a `test_reduce_scatter_post_commit.py` and `test_reduce_scatter_nightly.py`? Are we planning on breaking off directories as well by test type or not necessarily?

**Comment by SeanNijjar** [comment]
Note that post commit is clean but there are timeouts in the t3000 tests (not due to hangs but due to the length of the test suite). Going to work with @tapspatel to figure out where to put these tests.

File: tests/scripts/run_tests.sh (hunk 1 above)
Comment (tt-rkim) [reply tests/scripts/run_tests.sh]: The original sources can continue being placed according to current patterns, but we're planning to put them as symlinks to designate pipelines more clearly.

Take a look at `tests/nightly`, they symlink to a bunch of stuff in other places

**Review by tt-rkim — APPROVED** [review]
Run nightly for us, and this is good to go

**Comment by SeanNijjar** [comment]
> Run nightly for us, and this is good to go

Which nightly tests are you referring to?

**Comment by SeanNijjar** [comment]
> > Run nightly for us, and this is good to go
> 
> Which nightly tests are you referring to?

Cleared this up a little offline with @tt-rkim. The reduce scatter nightly invocation (and already present test_all_gather nightly invocation) shouldn't be in their current location in `run_frequent_api_pipeline_tests` because that doesn't run on t3000. Those invocations should be moved to the appropriate t3000 pipeline. 

At this point this means these tests are effectively disabled. For that reason I'm going to revert my change to add the reduce_scatter nightly invocation to  `run_frequent_api_pipeline_tests` and then open an issue to move it and test_all_gather nightly to the appropriate location. 

Okay with you @ttmchiou and @tt-rkim ?

**Review by ttmchiou — APPROVED** [review]
merge when post commit passes

**Pull request merged**